### ISSUE TITLE
Macros for making commands

### ISSUE THREAD

**Issue by summivox** [imxrt-rs/imxrt-dcd#1]
I am in the process of adding convenient macros for making commands, mirroring the RAL's `write_reg!` and friends.
Obviously, due to how DCD works, this can't be a 1:1 match:
- `modify_reg!` maps to two commands: clear masks, then set values
- `read_reg!` is loosely related to Check commands, but here we exclusively deal with masks / bitset operations.

I think the key difference here is that DCD often needs to explicitly deal with `field::mask`. To this end, I have a prototype that allows you to write something like:

```rust
vec![
    write_reg!(ral::ccm_analog, CCM_ANALOG, PLL_ARM, @POWERDOWN, DIV_SELECT: 75),
    check_all_set(ral::ccm_analog, CCM_ANALOG, PLL_ARM, @LOCK),
]
```

The syntax is shared across all commands: `{write,set,clear}_reg` and `check_{all,any}_{set,clear}`. Basically any argument following the register name can be:
- `@field` => equivalent to `field::mask`
- `field: value` => equivalent to the `(value << field::offset) & field::mask` (same as RAL)
- (arbitrary expression)

All arguments are OR'd together to form the final write value / check mask.

This is arguably a superset of RAL macros. However, I'm not sure about readability of the "@field" notation. Current candidates:
- sigil: `@field`, `#field`, ...
- key-value: `field: _`, `field: *`, ...
- Don't special case, because `field::mask` is already supported in this position.

cc @mciantyre --- what do you think?

**Comment by summivox** [comment]
Prototype pushed to https://github.com/imxrt-rs/imxrt-dcd/commit/36701a438baa962359173d50144bf0b577750446

(BTW: I found out that simultaneously setting `POWERDOWN` and `DIV_SELECT` doesn't work; nevertheless it serves the purpose of an example of capability).

**Comment by mciantyre** [comment]
Treating this as a superset of RAL macros works, and I like what's designed. If we wanted to disallow the sigil in `write_reg!`, we might be able to use field names as-is throughout all other macros. This would follow the example of [`reset_reg!`](https://docs.rs/ral-registers/0.1.3/ral_registers/macro.reset_reg.html), available in the upstream package.

```rust
// Before: '@field' is allowed
check_all_clear!(ral::ccm, CCM, CDHIPR, @PERIPH_CLK_SEL_BUSY, @PERIPH2_CLK_SEL_BUSY),

// After: '@field' replaced with just 'field' to look more like reset_reg!
check_all_clear!(ral::ccm, CCM, CDHIPR, PERIPH_CLK_SEL_BUSY, PERIPH2_CLK_SEL_BUSY),

// ---

// Before: '@field' permitted in write_reg!
write_reg!(ral::ccm_analog, CCM_ANALOG, PLL_ARM, @BYPASS, BYPASS_CLK_SRC: CLK1)

// After: '@field' not allowed in write_reg! and user should be explicit
write_reg!(ral::ccm_analog, CCM_ANALOG, PLL_ARM, BYPASS: 1, BYPASS_CLK_SRC: CLK1)
```

How do we handle the 1/2/4 bit width selection?

**Comment by summivox** [comment]
`reset_reg!` can get away without special markers because it is not designed to take arbitrary `value` expressions. For the DCD macros I would still want them to accept at least one arbitrary expression term, which would be ambiguous with a "naked" field name. IMHO all the macros here must at least support `..., field1: value1, field2: value2)` and `..., expr)`.

Regarding the bit width selection, considering the built-in address limitations, all the registers you are allowed to touch are 32-bit, so I don't see much value in enabling the 8-/16-bit options from the macro... That said, I think the RAL does encode the native bit width of the register, so maybe I can fish it out. I don't think it's a blocker though.

**Comment by summivox** [comment]
I somehow found out a way to coax out a width without making the impl too dirty --- `core::mem::size_of<Register>` works with some twisting. This logic can be hidden in plain sight as part of the `Width` enum.

https://github.com/imxrt-rs/imxrt-dcd/commit/d9cf0a02ddc5d290f1ca675660a308417ac42158

**Comment by mciantyre** [comment]
I hadn't considered allowing key-value syntax in the `check_*` macros. Here's a few examples I'm pondering. Are these the semantics of these macros?

`check_all_set!(ral::ccm_analog, CCM, CCM_ANALOG, PLL_ARM, LOCK: 1)`
- Semantics: "block until the PLL is locked in." Equivalent to the `@LOCK` syntax.

`check_all_set!(ral::ccm_analog, CCM, CCM_ANALOG, PLL_ARM, LOCK: 0)`
- Semantics: The user may have wanted to use `check_all_clear!`, or they should use `1` as the value for `LOCK`. This might decay to `(*address & 0) == 0`, always-true.

`check_all_set!(ral::ccm, CCM, CBCMR, PRE_PERIPH_CLK_SEL: 0b10)`
- Semantics: `PRE_PERIPH_CLK_SEL` is two bits wide. The command it generates is "bit 19 is set, so the clock selection is _either_ `PLL2_PFD0` or `PLL1` (assuming a 1060 MCU).

**Comment by summivox** [comment]
My understanding agrees with yours for all 3 examples (the instance needs to be provided though).

What I have in mind for the `field: value` syntax for both writes and checks:

- `read_reg!(..., FIELD == value)` maps to a two-command sequence (if it cannot change between the two checks):
   - `check_all_set!(ral::ccm, CCM, CBCMR, PRE_PERIPH_CLK_SEL: 0b10)`
   - `check_all_clear!(ral::ccm, CCM, CBCMR, PRE_PERIPH_CLK_SEL: !0b10)`

- `modify_reg!(..., FIELD: value)` maps to a two-command seqeuence (if there is no side effect triggered by writing twice):
  - `clear_reg!(..., @FIELD)`
  - `set_reg!(..., FIELD: value)`

- From the impl perspective, it's convenient to share the argument parsing logic across all macros in the family.

--------

I think a lot of the ambiguity comes from the word "set". What does it mean to have a field "set"? A user may reasonably think of:
- `read_reg!(..., FIELD == enumerator)`
- `read_reg!(..., FIELD == 0b111)`

This might prompt naming the check macros `check_{all,any}_{zero,one}` and renaming the enums accordingly. However:
- It would deviate from the terminology written in the RM (Table 9-45 of RT1060 RM rev 3), creating further cognitive overhead.
- Arguably, `check_any_zero` is still confusing because the mask is "negative logic", i.e. where the mask has a 1, the value needs to be a 0.

However, my counterpoints involve the target audience of this crate: someone building a BSP, who presumably/hopefully understands the subtle bitwise ops semantics here, and **reads the RM** (gasp). I don't want this crate to dictate what a DCD script author can or cannot do, beyond the most basic sanity checks. Linking to both the RAL and RM helps with readability and maintainability, but IMHO it's the user's responsibility to ensure that the caveats (esp. from the ROM bootloader) does not affect them. At the end of the day, each macro expands to a literal, and the user can always `assert_eq!` or even check the DCD binary output.

**Comment by summivox** [comment]
I spent a good chunk of my time writing DCD for setting up SEMC/SDRAM to get a feel on the user side. 

It actually felt quite ergonomic at this point. The `@FIELD` syntax has somewhat grown on me --- it's distinct enough but not too obnoxious despite being alien to Rust, with a nice side effect of visually separating register and field names so that I can tell from a glance.

I was reluctant releasing this because I would most likely be stuck with the decision ;) That said, given that it withstood the test of full PLL + clock tree + SEMC/SDRAM configuration, I think it should be ready after some more polish.

**Comment by summivox** [comment]
Published v1.1.0 .
